modin/pandas/dataframe.py

```diff
@@ -2700,6 +2700,118 @@ def style(df):
 
         return self._default_to_pandas(style)
 
+    def _build_batch_pipeline(self, func: Callable, axis: int, **kwargs):
```

Comment:
I think the use of methods starting with an underscore is not reliable for users. What if we would add a function named `build_batch_pipeline` and put it in `modin.pandas.utils`. The function would return a `Pipeline` object, which the user can operate on and add further queries to it.

Reply:
Or it might be better to have a function named `to_pipeline`/`to_modin_pipeline` by analogy to `to_pandas`. `to_pipeline` would return a `Pipeline` object, which the user can operate on.

Reply:
I am not sure we want to add a new API to `modin.pandas.DataFrame`, which should be identical to pandas and we are trying to match to. pandas DataFrame API is already numerous. Putting an additional API to `modin.pandas.DataFrame` will bloat this page [link1](https://modin--4286.org.readthedocs.build/en/4286/supported_apis/dataframe_supported.html) and this left sidebar for `modin.pandas.DataFrame` [link2](https://modin--4286.org.readthedocs.build/en/4286/flow/modin/pandas/dataframe.html), which we are trying to refine. It will be hard for users to find the pipeline API on those pages. I think it would be great to have a pipeline object and a separate documentaion page for it, where users will be able to easily discover what it is and how it works with. A `Pipeline` object would contain a `modin.pandas.DataFrame` object, which users can operate on. Does this make sense?

Reply:
As another option, we could create a child of `modin.pandas.DataFrame` using a function named `to_<child>`. That child would have a `Pipeline` member and stuff for building the pipeline. Users will be able to continue working with the child using both the pipeline API and `modin.pandas.DataFrame` API.

Reply:
I am rather for the latter case.

Reply:
I prefer the Pipeline object approach here over the `to_child` approach since it seems more natural to be to have `Pipeline` class over a `Pipeline` member. Agreed on separating the Batch API from the `modin.pandas.DataFrame`.

Reply:
I guess there is a misunderstanding here. Users won't be interacting with a pipeline member inside a dataframe directly, only through `add_query`, `compute_batch`, and other applicable stuff. Performing the `to_child` approach looks more tricky though. There might be the following case:

```python
df1 = df_child.add_query(...)
df2 = df1.add_query(...)
df3 = df2.abs() # or any other call of modin.pandas.DataFrame API
```

Calling `df2.abs()` make us drain all queries added before. So the Pipeline object approach looks simpler and probably more reasonable.

Reply:
I went ahead and just exposed the pipeline object (in line with @YarShev 's first suggestion). 

To use the Pipeline API, import the `PandasQueryPipeline` from `modin.experimental.batch.pipeline` and initialize it with a dataframe, an axis to pipeline over, and an optional number of partitions (defaults to `NPartitions.get`. Then, you can add queries using the Pipeline object's `add_query` method, change the dataframe to pipeline by using the `update_df` method, and compute the results using the `compute_batch` method. 

E.g.:
```python
import modin.pandas as pd
import numpy as np
from modin.experimental.batch.pipeline import PandasQueryPipeline
df = pd.DataFrame(np.random.randint(0, 1, (1000, 1000))
pipeline = PandasQueryPipeline(df, 0, num_partitions=20)
pipeline.add_query(lambda df: df * -3, is_output=True)
pipeline.update_df(pd.DataFrame([[1, 2, 3], [4, 5, 6]]))
outputs = pipeline.compute_batch()
```

Reply:
I am okay with this. Do you think it would be better to put a pipeline in `modin.experimental.pandas` to access it using something like this?

```python
from modin.experimental.pandas.batch.pipeline import QueryPipeline
# or
from modin.experimental.pandas.pipeline import QueryPipeline
```

@RehanSD, what do you expect to put near the `batch` folder in future?

Aside: I hope I will get to the code itself to look at 😄

Reply:
ping

Reply:
Oh sorry - totally missed this question! I'm not entirely sure to be honest. I wasn't sure if we planned to extend the `batch` API with other things (potentially utils or other methods of processing). I'm happy to change it, but I feel like `pipeline` alone with batch in the name is potentially confusing. Again, I'll defer to @devin-petersohn on what to do with this one!

Reply:
cc @RehanSD, @devin-petersohn, ping